Issue #40: Mesos now dynamically configures ports of the Node Manager

DarinJ opened the pull request:
Mesos now dynamically configures ports of the Node Manager.  This addresses Issue #40 completely and Issue #45 partially.

This is done by first checking if three ports are available in the resource offer, if so they are added the TaskInfo object and passed to the executor via NMTaskConfig. 

Added `NMPorts.java`, an immutable helper class used by `TaskFactory.java`, essentially a Tuple Object.

Tested on one node cluster, map reduce jobs complete as expected, all NM web ports are linked to the RM as expected.

Updated `myriad-dev.md` to reflect changes.  Other than the modifications to `yarn-site.xml` no other changes are needed.  If the modifications are left out, the node manage will use the default ports.

DarinJ commented:
I was planning on waiting until #45 was completely done, but the ports were really straight forward to add.  Also, due to the nature of mesos managing the port resources vs statically taking them, I think this change stands on it's own.

Comments welcome.

smarella commented on src/main/java/com/ebay/myriad/executor/NMTaskConfig.java:
```diff
@@ -11,6 +11,11 @@
+    private Long nmPort;
+    private Long nmWebAppPort;
+    private Long nmLocalizerPort;
+
```
MapReduce shuffle uses a http endpoint that needs to be added to this list.

smarella commented on src/main/java/com/ebay/myriad/scheduler/NMPorts.java:
```diff
@@ -0,0 +1,28 @@
+    private final long nmWebAppPort;
+    private final long nmLocalizerPort;
+    private final long nmPort;
+
```
shuffle http port needs to be added here as well.

smarella commented on src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java:
```diff
@@ -157,18 +167,21 @@ private boolean matches(Offer offer, NMProfile profile) {
                 && taskUtils.getAggregateMemory(profile) <= mem
                 && SchedulerUtils.isMatchSlaveAttributes(offer,
-                requestAttributes)) {
+                requestAttributes) && ports >= 3) {
```
please avoid using magic numbers in the code.

DarinJ replied:
This magic number appears in one other location as well.  There's a few other constants appearing in multiple files.  Is it worth adding a constants class to avoid changes to one and not another?

smarella replied:
It might be better to move the magic number into the relevant class to which it belongs. For e.g:

``` java
public class NMPorts {
  private static final String NM_RPC_PORT_KEY = "nm.rpc.port";
  private static final String NM_LOCALIZER_PORT_KEY = "nm.localizer.port";
  private static final String NM_WEBAPP_PORT_KEY = "nm.webapp.port";
  private static final String NM_HTTP_SHUFFLE_PORT_KEY = "nm.http.shuffle.port";

  private static final String NM_PORT_KEYS = 
                 new String[] {
                     NM_RPC_PORT_KEY,
                     NM_LOCALIZER_PORT_KEY,
                     NM_WEBAPP_PORT_KEY,
                     NM_HTTP_SHUFFLE_PORT_KEY
                  };

  private Map<String, Long> nmPorts = 
                  new HashMap<String, Long>(NM_PORT_KEYS.size());

  public NMPorts(long[] ports) { // takes an array rather than args
      if (ports.length != NM_PORT_KEYS.length) {
          // throw error 
      }
      for (int i = 0; i < ports.length; i++) {
         nmPorts.put(NM_PORT_KEYS[i], ports[i]);
      }
  }

  // the magic number is replaced by a static getter
  public static int getNumPorts() {
     return nmPorts.size();
  }

  public long getRpcPort() {
     return nmPorts.get(NM_RPC_PORT_KEY);
  }
  // getters for other ports 
}
```

DarinJ commented:
@smarella hopefully this addresses your concerns.  Let me know.

Forgot to mention in the commit but I also ensured multiple mappers were running so a shuffle involved more than one node.

DarinJ commented:
Going to have to fix some formatting, forgot to change a setting between projects (\* imports).  Will fix tomorrow.  Sorry for any inconvenience.

smarella commented on src/main/java/com/ebay/myriad/executor/NMTaskConfig.java:
```diff
@@ -11,6 +11,19 @@
+    private Long webAppHttpPort;
+    private Long shufflePort;
+
+    public Long getShufflePort() {
```
push the getter down to appear after all the fields.

smarella commented on src/main/java/com/ebay/myriad/executor/NMTaskConfig.java:
```diff
@@ -11,6 +11,19 @@
+        return shufflePort;
+    }
+
+    public void setShufflePort(Long shufflePort) {
```
push the setter down to appear after all the fields.

smarella commented on src/main/java/com/ebay/myriad/scheduler/TaskFactory.java:
```diff
@@ -176,6 +201,8 @@ public TaskInfo createTask(Offer offer, TaskID taskId, NodeTask nodeTask) {
             Objects.requireNonNull(nodeTask, "NodeTask should be non-null");
 
+            NMPorts ports = getPorts(offer);
+
```
Let's add a LOGGER.debug that prints out all the random ports assigned for each NM's end point. Perhaps we should just a toString() method inside NMPorts.java and do `LOGGER.debug(nmPorts.toString)`.

DarinJ replied:
Good idea

smarella commented on src/main/java/com/ebay/myriad/scheduler/TaskFactory.java:
```diff
@@ -57,6 +58,30 @@ public NMTaskFactoryImpl(MyriadConfiguration cfg, TaskUtils taskUtils) {
             this.taskUtils = taskUtils;
         }
 
+        //Utility function to get the first three ports of an offer
```
"three" in the comment is not true anymore

smarella commented on src/main/java/com/ebay/myriad/scheduler/TaskFactory.java:
```diff
@@ -57,6 +58,30 @@ public NMTaskFactoryImpl(MyriadConfiguration cfg, TaskUtils taskUtils) {
+            HashSet<Long> ports = new HashSet<Long>();
+            for (Resource resource : offer.getResourcesList()){
+                if (resource.getName().equals("ports")){
+                    Iterator<Value.Range> itr = resource.getRanges().getRangeList().iterator();
```
Why can't we simply pick `NMPorts.expectedNumPorts()` number of entries from `resource.getRanges().getRangeList()` list ? Like so:

``` java
private static NMPorts getPorts(Offer offer) {
  if (resource.getRanges().getRangeList().size() > NMPorts.getExpectedNumPorts()) {
     List<Long> ports = resource.getRanges().getRangeList().subList(0, NMPorts.getExpectedNumPorts());
     return new NMPorts(ports.toArray(new Long[0]));
  } else {
     LOGGER.warn("Offer doesn't contain enough ports to launch NM");
     return null; // the caller should abort launching NM if this happens
  }
}
```

DarinJ replied:
`getRanges.getRangeList()` returns a list of objects of type `Range`.  It has two methods `getBegin()` and `getEnd()`.  There could be exactly one `Range` in `getRanges.getRangeList` that would have 1000 ports.  This would cause the `if` statement to fail likewise `subList` wouldn't work as expected as the would be a sublist of `Range` not `Long`.  It may warrant a comment?

In general I expect to get one range, loop through the first 4 longs between `beginRange` and `endRange` and then be out.  But its possible due to other frameworks etc, that we get 500 ranges of size 1, in which case we loop through the the first 4 ranges.  Either way its bounded by 4.

smarella replied:
Ah.. I see the point. Thanks for clarifying.

smarella commented on src/main/java/com/ebay/myriad/scheduler/TaskFactory.java:
```diff
@@ -120,7 +145,7 @@ private String getConfigurationUrl() {
                 //The url for the resource manager config is: http(s)://hostname:port/conf so fetcher.cpp downloads the
                 //config file to conf, It's an xml file with the parameters of yarn-site.xml, core-site.xml and hdfs.xml.
-                String configCopyCmd = "cp conf " + cfg.getYarnEnvironment().get("YARN_HOME") +
+                String configCopyCmd = "sudo cp conf " + cfg.getYarnEnvironment().get("YARN_HOME") +
```
Why is the "sudo" added?

DarinJ replied:
Unrelated change related to diagnosing an issue with remote distribution.  Will change back and do a separate PR.

smarella commented on src/main/java/com/ebay/myriad/scheduler/TaskFactory.java:
```diff
@@ -57,6 +58,30 @@ public NMTaskFactoryImpl(MyriadConfiguration cfg, TaskUtils taskUtils) {
+            for (Resource resource : offer.getResourcesList()){
+                if (resource.getName().equals("ports")){
+                    Iterator<Value.Range> itr = resource.getRanges().getRangeList().iterator();
+                    while (itr.hasNext() && ports.size() < 3) {
```
`ports.size() < 3` has now changed.

smarella commented on docs/myriad-dev.md:
````diff
@@ -72,6 +90,17 @@ export MESOS_NATIVE_JAVA_LIBRARY=/usr/local/lib/libmesos.so
+```xml
+<!--This option enable dynamic port assignment by mesos -->
+<property>
+    <name>mapred.shuffle.port</name>
````
The property is "mapreduce.shuffle.port" (as listed [here](https://hadoop.apache.org/docs/r2.7.0/hadoop-mapreduce-client/hadoop-mapreduce-client-core/mapred-default.xml))

smarella commented on src/main/java/com/ebay/myriad/executor/MyriadExecutor.java:
```diff
@@ -63,6 +62,16 @@
+
+    public static final String KEY_NM_WEBAPP_ADDRESS = "nodemanager.webapp.address";
+
+    public static final String KEY_NM_SHUFFLE_PORT = "myriad.shuffle.port";
```
Let's go with "myriad.mapreduce.shuffle.port"

smarella commented on docs/myriad-dev.md:
```diff
@@ -72,6 +90,17 @@ export MESOS_NATIVE_JAVA_LIBRARY=/usr/local/lib/libmesos.so
+<!--This option enable dynamic port assignment by mesos -->
+<property>
+    <name>mapred.shuffle.port</name>
+    <value>${myiad.shuffle.port}</value>
```
type - should be "myriad".

Also, as mentioned in another comment, let's have the property named to "myriad.mapreduce.shuffle.port"

DarinJ replied:
Related should we put myriad in front of all of these including  myriad.nodemanager.resource.memory-mb etc?  I can do all the ports in this PR.  Probably separate for cpu and mem?

smarella replied:
+1 for that. It really helps to distinctly tell which YARN properties are myriad setting the values for.

DarinJ commented:
Other than the extended comment I think this can get done by COB tomorrow.  Thanks and sorry for the sloppy items.

DarinJ commented:
@smarella finished changes.

kensipe commented:
please rebase and I'll take a look

DarinJ commented:
Holding off rebase until I look over FGS.

pdread100 commented on src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java:
```diff
@@ -125,7 +126,7 @@ public void onEvent(ResourceOffersEvent event, long sequence,
     private boolean matches(Offer offer, NMProfile profile) {
         double cpus = -1;
```
As a general comment, and even though we do not use the PMD codesize option (maybe we should) this method has exceeded the accepted standards for Cyclomatic Complexity. The method matches() has a complexity of 18 and a  NpathComplexity of over 200. The general accepted CC level is 5-7, though some feel 10 should be the limit. The output of the codesize PMD option looks like:

9 myriad/src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java     50  The class 'ResourceOffersEventHandler' has a Cyclomatic Complexity of 14 (Highest = 18).
10  myriad/src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java   126     The method 'matches' has a Cyclomatic Complexity of 18.
11  myriad/src/main/java/com/ebay/myriad/scheduler/event/handlers/ResourceOffersEventHandler.java   126     The method matches() has an NPath complexity of 440

I took the liberty of refactoring the code as an example on how to reduce the complexity, and remove the PMD issues. Though this increased the overall complexity of the class it reduced the complexity per method which would simplify, in theory, unit test, code coverage, and the general quality of the code. The following is the refactor:

```
 private boolean matches(Offer offer, NMProfile profile) {
        Map<String, Object> results = new HashMap<String, Object>(5);

        for (Resource resource : offer.getResourcesList()) {
            if (resourceEvaluators.containsKey(resource.getName())) {
                resourceEvaluators.get(resource.getName()).eval(resource, results);
            } else {
                LOGGER.warn("Ignoring unknown resource type: {}",
                        resource.getName());
            }
        }
        double cpus = (Double) results.get("cpus");
        double mem = (Double) results.get("mem");
        int ports = (Integer) results.get("ports");

        checkResource(cpus < 0, "cpus");
        checkResource(mem < 0, "mem");
        checkResource(ports < 0, "port");

        return checkAggregates(offer, profile, ports, cpus, mem);
    }

    private void checkResource(boolean fail, String resource) {
        if (fail) {
            LOGGER.error("No " + resource + " resources present");
        }
    }

    private boolean checkAggregates(Offer offer, NMProfile profile, int ports, double cpus, double mem) {
        Map<String, String> requestAttributes = new HashMap<>();

        if (taskUtils.getAggregateCpus(profile) <= cpus
                && taskUtils.getAggregateMemory(profile) <= mem
                && SchedulerUtils.isMatchSlaveAttributes(offer, requestAttributes)
                && NMPorts.expectedNumPorts() <= ports) {
            return true;
        } else {
            LOGGER.info("Offer not sufficient for task with, cpu: {}, memory: {}, ports: {}",
                    taskUtils.getAggregateCpus(profile), taskUtils.getAggregateMemory(profile), ports);
            return false;
        }      
    }

    private static Double scalarToDouble(Resource resource, String id) {
        Double value = new Double(0.0);
        if (resource.getType().equals(Value.Type.SCALAR)) {
            value = new Double(resource.getScalar().getValue());
        } else {
            LOGGER.error(id + " resource was not a scalar: {}", resource
                    .getType().toString());
        }
        return value;
    }

    private interface EvalResources {
        public void eval(Resource resource, Map<String, Object>results);
    }

    private static Map<String, EvalResources> resourceEvaluators;

    static {
        resourceEvaluators = new HashMap<String, EvalResources>(4);
        resourceEvaluators.put("cpus", new EvalResources() {
            public void eval(Resource resource, Map<String, Object> results) {
                results.put("cpus", scalarToDouble(resource, "cpus"));
            }
        });
        resourceEvaluators.put("mem", new EvalResources() {
            public void eval(Resource resource, Map<String, Object> results) {
                results.put("mem", scalarToDouble(resource, "mem"));
            }
        });
        resourceEvaluators.put("disk", new EvalResources() {
            public void eval(Resource resource, Map<String, Object> results) {
            }
        }); 
        resourceEvaluators.put("ports", new EvalResources() {
            public void eval(Resource resource, Map<String, Object> results) {
                int ports = 0;
                if (resource.getType().equals(Value.Type.RANGES)) {
                    Value.Ranges ranges =  resource.getRanges();
                    for (Value.Range range : ranges.getRangeList()) {
                        if (range.getBegin() < range.getEnd()) {
                            ports += range.getEnd() - range.getBegin() + 1;
                        }
                    }

                } else {
                    LOGGER.error("ports resource was not Ranges: {}", resource
                            .getType().toString());

                }
                results.put("ports", Integer.valueOf(ports));
            }
        });

    }

```

Also I will note that using String compare in else if is very slow. By switching to a HashMap you get a bump in speed.

DarinJ commented:
Went ahead rebased and added @pdread100 suggestions.  @smarella I tested on a 4 node cluster, all went well.

smarella commented:
Tested on a 4 node cluster on MapR distribution. Verified NM uses random ports for various end points. Worked pretty well.  Merging in the PR.
